Introduce Continuum Transfunctioner daemon

marcopeereboom opened the pull request:
**Summary**
Add a new daemon that handles Threshold Signature Scheme(TSS) magic for MinerFI.

This is being separated into several PRs to keep them smaller and more reviewed.

In order for this to move to production we must at least add:
* Spec for continuum protocol
* HIPPO for transfunctioner

marcopeereboom committed:
Introduce Continuum Transfunctioner daemon

marcopeereboom committed:
backup this test code

marcopeereboom committed:
back this up

marcopeereboom committed:
Make nonce counter an atomic

marcopeereboom committed:
Handshake

marcopeereboom committed:
Add transport close and fix some panics

marcopeereboom committed:
Work around several linter items

marcopeereboom committed:
Remove some leftovers

marcopeereboom committed:
remove sleep with a bunch of scary looking code

marcopeereboom committed:
tagline

marcopeereboom committed:
Make dns server a bit easier to use and prep test framework for it

marcopeereboom committed:
Verify all dns records and proper lookup

marcopeereboom committed:
Print env variable name when failing

marcopeereboom committed:
Use signal stuff joshua likes

marcopeereboom committed:
Add some nuggets for antonio and some cleanup

AL-CT committed:
change key exchange to match keys & add testing

marcopeereboom committed:
Preliminary DNS verification

marcopeereboom committed:
Move dnsAppName to protocol

AL-CT committed:
add note to test

AL-CT committed:
add session management to continuum

AL-CT committed:
add some configs and block connections when max is reached

marcopeereboom committed:
Fixup DNS a bit

marcopeereboom committed:
kinda working dns stuff

AL-CT committed:
for marco

marcopeereboom committed:
listen on a port that isnt 127.0.0.1

AL-CT committed:
fix DNS handshake test

AL-CT committed:
change nomenclature in tests

marcopeereboom committed:
backup before doing more drastic diags

marcopeereboom committed:
show antonio

marcopeereboom committed:
Clip size bytes

marcopeereboom committed:
missed test file

AL-CT committed:
apply marco's fix

AL-CT committed:
also fix DNS test

marcopeereboom committed:
order for wire order

marcopeereboom committed:
Close connection on the way out; think about this some mroe

marcopeereboom committed:
linter stuff

marcopeereboom committed:
Remove debug and make at correct

marcopeereboom committed:
bit more cleanup and common debug

marcopeereboom committed:
write and check n as god intended

marcopeereboom committed:
add timeoout to write too

marcopeereboom committed:
sprinkle moar debag

marcopeereboom committed:
run more than once

marcopeereboom committed:
run positive path a bunch to show antonio his code is broken ;-P

marcopeereboom committed:
This seemd to provoke some timeoouts

marcopeereboom committed:
oops

AL-CT committed:
separate client and server

marcopeereboom committed:
this kinda works

marcopeereboom committed:
Check identities

marcopeereboom committed:
bring back transport encryption/decryption test

marcopeereboom committed:
Start adding some doco

marcopeereboom committed:
Add more doco and pretty prints

marcopeereboom committed:
Some more stuff

AL-CT committed:
test all curves

marcopeereboom committed:
Add DNS verification

marcopeereboom committed:
show data race avoidance code

marcopeereboom committed:
missing doco

marcopeereboom committed:
linter things

marcopeereboom committed:
Cleanup

marcopeereboom committed:
more linter

AL-CT committed:
add more docu

AL-CT committed:
clean up top level descriptions

AL-CT committed:
test all curves

marcopeereboom committed:
unfuck dns test

AL-CT committed:
use free ports

AL-CT committed:
key exchange errors

marcopeereboom committed:
oops forgot to commit ping

marcopeereboom committed:
remove dead code and set some sane defaults

marcopeereboom committed:
wrap size read also in a loop

marcopeereboom committed:
Test PayloadHash

marcopeereboom committed:
negative test as well

marcopeereboom committed:
grammar son

marcopeereboom committed:
Fix couple of XXX

marcopeereboom committed:
Timeouts

marcopeereboom committed:
Unused function

marcopeereboom committed:
Linter

marcopeereboom committed:
github complaint

AL-CT committed:
test handshake errors

marcopeereboom committed:
Test nonce

AL-CT committed:
test secret and identity

marcopeereboom committed:
WHAT THE FUCK

marcopeereboom committed:
...

AL-CT committed:
fix comflicting test names

marcopeereboom committed:
Derp have to travel identity during hello

marcopeereboom committed:
Doco

marcopeereboom committed:
debug panic

AL-CT committed:
add test cases for mismatched identity

marcopeereboom committed:
Test TXT record some more

marcopeereboom committed:
linter

marcopeereboom committed:
XXX nugget

joshuasing commented on service/continuum/continuum.go:
```diff
@@ -0,0 +1,395 @@
+		id := genID()
+		s.mtx.Lock()
+		if _, ok := s.sessions[id]; ok {
+			// ID is already used, retry.
```
bug: `s.mtx` is not unlocked before `continue`, resulting in a deadlock.
```suggestion
			// ID is already used, retry.
			s.mtx.Unlock()
```

marcopeereboom committed:
trfd->transfunctionerd, from @joshua

marcopeereboom commented on service/continuum/continuum.go:
```diff
@@ -0,0 +1,395 @@
+	Secret                  string
+	ListenAddress           string
+	MaxConnections          int
+}
```
Fixed.

marcopeereboom commented on service/continuum/continuum.go:
```diff
@@ -0,0 +1,395 @@
+// Use of this source code is governed by the MIT License,
+// which can be found in the LICENSE file.
+
+package continuum
```
Fixed

marcopeereboom committed:
Various nits from @joshua and a deadlockservice/continuum/protocol.go

marcopeereboom committed:
Make TXT record use ; per DKIM, from @joshua

marcopeereboom commented on service/continuum/protocol.go:
```diff
@@ -0,0 +1,954 @@
+		m[kv[0]] = kv[1]
+	}
+	return m, nil
+}
```
This was a crapshoot tbh. Other record's do not use no `;`. I picked it because less code really but it does make sense to me. Fixed.

marcopeereboom commented on service/continuum/continuum.go:
```diff
@@ -0,0 +1,395 @@
+// Use of this source code is governed by the MIT License,
+// which can be found in the LICENSE file.
+
+package continuum
```
Well yes but in this case the protocol is called `continuum` and the daemon is `transfunctionerd`. I do expect that `continuum` will be picked up by external parties that will run their own version. Let me have me my joke :-)

marcopeereboom committed:
Fix potential crash with improper sized blob, pointed out by @joshua

marcopeereboom committed:
Fix potential crash with improper sized blob, pointed out by @joshua

AL-CT committed:
add docu; limit message size

AL-CT committed:
reduce nonce test size for ci

marcopeereboom committed:
Fix potential unknown key-share vector.

@joshua correctly pointed out that I missed including part of the
transport keys as part of the message that must be signed. Modify code
to include ephemeral transport public key in the challenge that is to be
signed.

marcopeereboom committed:
doco

marcopeereboom committed:
Flesh out protocol doco

marcopeereboom commented on cmd/transfunctionerd/transfunctionerd.go:
```diff
@@ -0,0 +1,142 @@
+			Help:         "data directory",
+			Print:        config.PrintAll,
+		},
+		"TRF_SECRET": config.Config{
```
Fixed

marcopeereboom committed:
Nits from @joshuasing

marcopeereboom committed:
juju v2

marcopeereboom commented on go.mod:
```diff
@@ -1,22 +1,25 @@
 	github.com/dustin/go-humanize v1.0.1
 	github.com/ethereum/go-ethereum v1.16.7
 	github.com/go-test/deep v1.1.1
+	github.com/juju/loggo v1.0.0
```
Fixed everywhere.

marcopeereboom commented on go.mod:
```diff
@@ -1,22 +1,25 @@
 module github.com/hemilabs/heminetwork/v2
 
-go 1.25
+go 1.25.0
```
I have to stress again that this is done by the linter and/or gofumpt..

marcopeereboom committed:
tidy

marcopeereboom committed:
Add changelog
